packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx

```diff
@@ -0,0 +1,580 @@
+        .filter((item, filteredIndex, items) => {
+          return (
+            filteredIndex === 0 ||
+            item.type === SwatchColorPickerItemType.MenuItem ||
+            (item.index - items[filteredIndex - 1].index !== 1)
+          );
+        }));
+  }
+
+  /**
+   * Render the separator
+   * @param item - The divider item to get the data to render from
+   * @returns {JSX.Element} - Element that represents the separator
+   */
+  private _renderSeparator(item: ISwatchColorPickerItemProps): JSX.Element {
+    return this._renderHeaderOrDivider(item, 'separator', styles.divider);
+  }
+
+  /**
+   * Render the header
+   * @param item - The header item to get the data to render from
+   * @returns {JSX.Element} - Element that represents the header
+   */
+  private _renderHeader(item: ISwatchColorPickerItemProps): JSX.Element {
+    return this._renderHeaderOrDivider(item, 'heading', styles.header);
+  }
+
+  /**
+   * Handle the rendering of the header/divider
+   * @param item - The item to get the data to render from
+   * @param role - The role of the element
+   * @param className - the className to use
+   * @returns {JSX.Element} - Element that represents the header
+   */
+  private _renderHeaderOrDivider(item: ISwatchColorPickerItemProps, role: string, className: string): JSX.Element {
+    return <div
+      role={ role }
+      key={ item.id }
+      className={ className }
+    >
+      { item.label && item.label }
+    </div>;
+  }
+
+  /**
+   * Render a cell or menu item
+   * @param item - The item to render
+   * @param posInSet - Optinal, the position in the set of the item
+   * @param setSize - Optional, the total set size this item is in
+   * @returns {JSX.Element} - Element representing the item
+   */
+  @autobind
+  private _renderOption(item: ISwatchColorPickerItemProps, posInSet: number = null, setSize: number = null): JSX.Element {
+    let id = this._id;
+    let isCell = item.type === SwatchColorPickerItemType.Cell;
+    return (
+      <CommandButton
+        { ...item.menuItemButtonProps }
+        id={ id + '-list' + item.index }
+        ref={ 'option' + item.index }
+        key={ id + item.id }
+        data-index={ item.index }
+        data-is-focusable={ true }
+        aria-posinset={ !isCell ? (posInSet && posInSet) : null }
+        aria-setsize={ !isCell ? (setSize && setSize) : null }
+        disabled={ this.props.disabled || item.disabled }
+        className={ css(
+          (isCell ? styles.cell : styles.item),
+          {
+            ['is-selected ' + styles.cellIsSelected]: (isCell && this.state.selectedIndex === item.index),
+            ['is-disabled ' + styles.disabled]: item.disabled && item.disabled
+          },
+        ) }
+        onClick={ () => isCell ? this._onCellClick(item) : this._onMenuItemClick(item) }
+        onMouseEnter={ isCell ? () => this._onItemHoverOrFocused(item, this.props.onCellHovered) : null }
```

Comment:
You can create a new component for an option. It would take the item in as a prop and you can access it that way in your callbacks. I usually find arrow functions in render functions as a good signal that a component should be extracted at a particular layer. Any concerns against creating a component here?